### PR TITLE
require 'static on LocalExecutor::spawn, matching spawn_local

### PR DESCRIPTION
**Pull request by utilitydelta**
Without static its possible to borrow a local, spawn, detach, drop the local, then run(). The task then reads freed memory.

Regression from https://github.com/glommio/glommio/pull/30

### CODE REVIEW

File: glommio/src/executor/mod.rs [hunk 1]
```diff
@@ -1372,7 +1391,10 @@ impl LocalExecutor {
     ///
     /// [`spawn_local`]: crate::spawn_local
     /// [`run`]: LocalExecutor::run
-    pub fn spawn<T>(&self, future: impl Future<Output = T>) -> Task<T> {
+    pub fn spawn<T>(&self, future: impl Future<Output = T> + 'static) -> Task<T>
```
Comment (dahankzter) [inline glommio/src/executor/mod.rs:1394]: Nice and I should have caught this but thanks for fixing it @utilitydelta !

File: glommio/tests/spawn_public.rs [hunk 2]
```diff
@@ -1,102 +1,66 @@
...
     });
-
-    assert_eq!(result, 42);
+    assert_eq!(executor.run(task), 42);
 }
 
-// This test should NOT compile if uncommented - demonstrates !Send enforcement
-/*
 #[test]
-fn test_cannot_send_executor_between_threads() {
```
Comment (dahankzter) [inline glommio/tests/spawn_public.rs:75]: Maybe we can add a compile_fail doctest for this?

**Review by dahankzter — APPROVED** [review]
Nice work I have just a suggestion about a doctest but this looks good to me.

File: glommio/tests/spawn_public.rs [hunk 3]
```diff
@@ -1,102 +1,66 @@
...
+                futures_lite::future::yield_now().await;
+                seen.set(payload[0]);
+            })
+            .detach();
+    }
+    executor.run(async {
+        for _ in 0..8 {
+            futures_lite::future::yield_now().await;
+        }
     });
```
Comment (vlovich) [inline glommio/tests/spawn_public.rs:64]: Can you please help me understand the purpose of the yield now + the magic 0..8 loop? Where does the 8 come from? Would it be perhaps clearer to use channels as a synchronization mechanism? Is there even some need for having two tasks in the first place?

File: glommio/tests/spawn_public.rs [hunk 4]
```diff
@@ -1,102 +1,66 @@
...
-    // This should fail to compile with:
-    // error[E0277]: `Rc<RefCell<ExecutorQueues>>` cannot be sent between threads safely
-    std::thread::spawn(move || {
-        executor.spawn(async { 42 });
-    });
+    let task;
+    {
+        let local = String::from("moved into the task");
+        task = executor.spawn(async move { local });
+    } // local is gone, the task owns it
```
Comment (vlovich) [inline glommio/tests/spawn_public.rs:42]: Local is gone due to the closure move capture, not the end of the scope.

File: glommio/tests/spawn_public.rs [hunk 5]
```diff
@@ -1,102 +1,66 @@
...
-        sum
-    });
-
-    let result = executor.run(task);
-
-    assert_eq!(result, 5050); // Sum of 1 to 100
-}
-
-#[test]
-fn test_spawn_inside_run_still_works() {
```
Comment (vlovich) [inline glommio/tests/spawn_public.rs:58]: why the removal of this test?

File: glommio/src/executor/mod.rs [hunk 6]
```diff
@@ -1372,7 +1391,10 @@ impl LocalExecutor {
...
+    where
+        T: 'static,
```
Comment (vlovich) [inline glommio/src/executor/mod.rs:1396]: This is the blocker for me in this PR: this is a safe function calling a safe function that apparently can result in unsafe behavior. To me that indicates there's a lifetime missing on an unsafe function somewhere that needs to be fixed.

**Review by vlovich — COMMENTED** [review]
Good find, but I don't believe this fixes the underlying problem. I'd like to see the doctest to make sure there's a compile error enforcing the 'static, but more importantly I'd like to see the lifetime fixed more thoroughly - a safe function calling safe functions with the wrong lifetime isn't possible. This implies there's an unsafe function or block somewhere that doesn't require the right lifetime on its arguments.

**Comment by utilitydelta** [comment]
good feedback guys. There is more scope to this, quite a few fn's should be marked as unsafe due to the future: F in spawn being swallowed in RawTask::allocate, and the compiler can't track the lifetime rules after that. So there are a bunch of functions that need unsafe markers. It might take me a day or two to properly make the right changes
